### PR TITLE
Rework trait and weapon docs page generation

### PR DESCRIPTION
**Pull request by penev92**
Initially I just wanted a flag to export the weapon docs info as JSON so that [the OpenRA Language Server (oraide)](https://github.com/penev92/Oraide.LanguageServer/) can use it in certain scenarios, but after a request by @pchote [on Discord](https://discord.com/channels/153649279762694144/388282819371204608/951233132215013457) I landed on the current solution of switching all exports to JSON and having a Python script to generate Markdown.
 - The first commit is a straight up improvement of the current code and needs to get in no matter what.
~~- After this PR is done, I'll need to do the same thing for the traits docs. (I also have the same thing lined up for sequences documentation)~~

There are two goals here:
 - To slightly update https://docs.openra.net/en/latest/release/weapons/ (without introducing any major changes). You can see the generated Markdown (generated from the generated JSON) here: https://github.com/penev92/OpenRA/wiki/Weapon-properties-documentation.
 - Same for https://docs.openra.net/en/latest/release/traits with the new Markdown showcased here: https://github.com/penev92/OpenRA/wiki/Trait-documentation-example.
 - The JSON will also be used by [oraide](https://github.com/penev92/Oraide.LanguageServer/).

P.S.: The final result is now also available on a test version of the docs site:
https://docs.openra.net/en/test/dev/traits/
https://docs.openra.net/en/test/dev/weapons/

### CODE REVIEW

File: OpenRA.Mods.Common/UtilityCommands/ExtractWeaponDocsCommand.cs [hunk 1]
```diff
@@ -36,66 +37,54 @@ void IUtilityCommand.Run(Utility utility, string[] args)
...
+					{
+						Name = fi.YamlName,
+						DefaultValue = FieldSaver.SaveField(objectCreator.CreateBasic(x.Type), fi.Field.Name).Value.Value,
+						Type = Util.FriendlyTypeName(fi.Field.FieldType),
+						Description = string.Join(" ", fi.Field.GetCustomAttributes<DescAttribute>(true).SelectMany(d => d.Lines)),
+						OtherAttributes = fi.Field.CustomAttributes
+							.Where(a => a.AttributeType.Name != "DescAttribute")
+							.Select(a => new
+							{
+								Name = a.AttributeType.Name.Substring(0, a.AttributeType.Name.Length - 9),
```
Comment (RoosterDragon) [inline OpenRA.Mods.Common/UtilityCommands/ExtractWeaponDocsCommand.cs]: Consider an `EndsWith` guard here in case there's an attribute not following the convention of using an `Attribute` suffix.

**Review by RoosterDragon — DISMISSED** [review]

**Review by RoosterDragon — DISMISSED** [review]

**Comment by penev92** [comment]
Fixed a couple more small nits plus the above comment.
Also updated the PR description.

File: OpenRA.Mods.Common/UtilityCommands/ExtractWeaponDocsCommand.cs [hunk 2]
```diff
@@ -36,66 +37,60 @@ void IUtilityCommand.Run(Utility utility, string[] args)
...
+					Description = string.Join(" ", x.Type.GetCustomAttributes<DescAttribute>(false).SelectMany(d => d.Lines)),
+					InheritedTypes = x.Type.BaseTypes().Select(y => y.Name).Where(y => y != x.Name && y != $"{x.Name}Info" && y != "Object").ToArray(),
+					Properties = FieldLoader.GetTypeLoadInfo(x.Type).Select(fi => new
+					{
+						Name = fi.YamlName,
+						DefaultValue = FieldSaver.SaveField(objectCreator.CreateBasic(x.Type), fi.Field.Name).Value.Value,
+						Type = Util.FriendlyTypeName(fi.Field.FieldType),
+						Description = string.Join(" ", fi.Field.GetCustomAttributes<DescAttribute>(true).SelectMany(d => d.Lines)),
+						OtherAttributes = fi.Field.CustomAttributes
+							.Where(a => a.AttributeType.Name != nameof(DescAttribute))
```
Comment (RoosterDragon) [inline OpenRA.Mods.Common/UtilityCommands/ExtractWeaponDocsCommand.cs]: Maybe also exclude `LoadUsing`?

Comment (penev92) [reply OpenRA.Mods.Common/UtilityCommands/ExtractWeaponDocsCommand.cs]: You may be right.

File: packaging/docs-weapons.py [hunk 3]
```diff
@@ -0,0 +1,50 @@
+import argparse
+import json
+from collections import OrderedDict
+
+if __name__ == "__main__":
+    parser = argparse.ArgumentParser()
+    parser.add_argument('--input', type=str, required=True)
```
Comment (pchote) [inline packaging/docs-weapons.py]: Remove the `--` here and in the line below, then call this as `python3 ./packaging/docs-weapons.py docs/api/release/weapons.json docs/api/release/weapons.md`

File: packaging/docs-weapons.py [hunk 4]
```diff
@@ -0,0 +1,50 @@
+import argparse
```
Comment (pchote) [inline packaging/docs-weapons.py]: License header.

File: packaging/docs-weapons.py [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+    parser.add_argument('--output', type=str, required=True)
+    args = parser.parse_args()
+
+    inputFile = open(args.input, "r")
+    jj = json.loads(inputFile.read())
+    inputFile.close()
+
+    weaponsByNamespace = OrderedDict()
+    for currentWeapon in jj["WeaponTypes"]:
+        if currentWeapon["Namespace"] in weaponsByNamespace: weaponsByNamespace[currentWeapon["Namespace"]].append(currentWeapon)
```
Comment (pchote) [inline packaging/docs-weapons.py]: Newline after `:` here and below.

File: packaging/docs-weapons.py [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+        else: weaponsByNamespace[currentWeapon["Namespace"]] = [ currentWeapon ]
```
Comment (pchote) [inline packaging/docs-weapons.py]: No spaces between `[ ]`

File: packaging/docs-weapons.py [hunk 7]
```diff
@@ -0,0 +1,50 @@
...
+    outputFile = open(args.output, "w")
+    outputFile.write("#### This documentation is aimed at modders. It displays a template for weapon definitions " +
+				"as well as its contained types (warheads and projectiles) with default values and developer commentary. " +
+				"Please do not edit it directly, but add new `[Desc(\"String\")]` tags to the source code.\n\n")
+
+    outputFile.write(f"This file has been automatically generated for version **{version}** of OpenRA.")
+    outputFile.write('\n\n#\n\n')
+
+    for namespace in weaponsByNamespace:
+        outputFile.write(f'## {namespace}\n')
```
Comment (pchote) [inline packaging/docs-weapons.py]: `outputFile.writeline` and drop the explicit `\n`'s.

File: packaging/docs-weapons.py [hunk 8]
```diff
@@ -0,0 +1,50 @@
...
+        outputFile.write('\n')
+
+        for weapon in weaponsByNamespace[namespace]:
+            outputFile.write(f'### {weapon["Name"]}\n')
+            outputFile.write(f'#### {weapon["Description"]}\n')
+            if weapon["InheritedTypes"]: outputFile.write("Inherits from: " + ", ".join([f'`{x}`' for x in weapon["InheritedTypes"]]) + '\n')
```
Comment (pchote) [inline packaging/docs-weapons.py]: Newline after `:` and `writeline`.

**Review by RoosterDragon — DISMISSED** [review]

File: packaging/docs-weapons.py (hunk 7 above)
Comment (penev92) [reply packaging/docs-weapons.py]: My google search found no evidence that there is a `.writeline` method :?
I found `.writelines`, which I guess I could use 🤷‍♂️

**Comment by penev92** [comment]
Rebased.

**Comment by penev92** [comment]
So I reworked the Python script to be able to generate documentation Markdown files for traits *and* weapons (and sequences) instead of having separate scripts. I also added the trait JSON generation here.
In this latest turn of events #19986 will now depend on this and will add everything sequence-related when this is merged.

File: OpenRA.Mods.Common/UtilityCommands/ExtractTraitDocsCommand.cs [hunk 9]
```diff
@@ -35,74 +36,58 @@ void IUtilityCommand.Run(Utility utility, string[] args)
...
+				});
+
+			var result = new
+			{
+				Version = version,
+				TraitInfos = traitTypesInfo
+			};
 
-			Console.Write(doc.ToString());
+			return Newtonsoft.Json.JsonConvert.SerializeObject(result);
```
Comment (pchote) [inline OpenRA.Mods.Common/UtilityCommands/ExtractTraitDocsCommand.cs]: Can we please have a `using Newtonsoft.Json` above and then `return JsonConvert.SerializeObject(result, Formatting.Indented);` for these?

File: packaging/generate-docs.py [hunk 10]
```diff
@@ -0,0 +1,82 @@
...
+                        outputFile.write(f'| {prop["PropertyName"]} | {prop["DefaultValue"]} | {prop["Type"]} | {attributes} | {prop["Description"]} |\n')
+                    else:
+                        outputFile.write(f'| {prop["PropertyName"]} | {prop["DefaultValue"]} | {prop["Type"]} | {prop["Description"]} |\n')
+
+            outputFile.write('\n#\n\n')
+
+    outputFile.close()
+
+if __name__ == "__main__":
+    parser = argparse.ArgumentParser()
```
Comment (pchote) [inline packaging/generate-docs.py]: IMO it would be cleaner if we called this file `format-docs.py` and had it read directly from stdin (`json.load(sys.stdin)`) and write directly to stdout (`print()`) instead of using temporary files. Can also add `#!/usr/bin/env python3` to the top of the script and mark the file as executable (`git update-index --chmod=+x packaging/format-docs.py`).

The workflow calls would then simplify to e.g.

```bash
./utility.sh all --docs "${GIT_TAG}" | ./packaging/format-docs.py > docs/api/playtest/traits.md
```

Comment (penev92) [reply packaging/generate-docs.py]: Should be done now.

**Comment by penev92** [comment]
Updated to what I hope is its final form.

**Comment by anvilvapre** [comment]
It was not your intention to sort the namespaces and types alphabetically? Only to list them order of appearance?

**Comment by penev92** [comment]
I'm not sure how to read the question, but I don't think there should be a change in the order - they are ordered by namespace, just like before.

**Comment by anvilvapre** [comment]
> I'm not sure how to read the question, but I don't think there should be a change in the order - they are ordered by namespace, just like before.

Because it's documentation I thought perhaps you intended to order namespaces alphabetically - but you currently don't. Perhaps the namespaces are already sorted when you read them from stdin.

**Comment by penev92** [comment]
```cs
var traitInfos = objectCreator.GetTypesImplementing<TraitInfo>().OrderBy(t => t.Namespace);
```
and
```cs
var warheads = objectCreator.GetTypesImplementing<IWarhead>().OrderBy(t => t.Namespace);
var projectiles = objectCreator.GetTypesImplementing<IProjectileInfo>().OrderBy(t => t.Namespace);
```
ensure things are ordered by namespace alphabetically before we even start.

**Comment by anvilvapre** [comment]
right. i only read the python script. sorry for the noise.

**Comment by penev92** [comment]
Rebased to fix a merge conflict.

**Comment by PunkPun** [comment]
`ExtractWeaponDocsCommand.cs` is being changed and reworked in 3 separate commits. It would be good if you could squash them. Or at least squash the changes to this file

File: packaging/format-docs.py [hunk 11]
```diff
@@ -0,0 +1,82 @@
...
+                        if not attributesList:
+                            attributes = ''
+                        else:
+                            attributes = ", ".join(attributesList)
+
+                        print(f'| {prop["PropertyName"]} | {prop["DefaultValue"]} | {prop["UserFriendlyType"]} | {attributes} | {prop["Description"]} |')
+                    else:
+                        print(f'| {prop["PropertyName"]} | {prop["DefaultValue"]} | {prop["UserFriendlyType"]} | {prop["Description"]} |')
+
+            print('\n#\n')
```
Comment (PunkPun) [inline packaging/format-docs.py]: Why is this needed? Now after every entry a hash is placed. To match the old generation remove this line entirely

File: packaging/format-docs.py [hunk 12]
```diff
@@ -0,0 +1,82 @@
...
+    print(f"This documentation is aimed at modders and has been automatically generated for version `{version}` of OpenRA. " +
+				"Please do not edit it directly, but instead add new `[Desc(\"String\")]` tags to the source code.\n")
+
+    print(f"Listed below are {explanation}:\n")
+
+    for namespace in typesByNamespace:
+        print(f'## {namespace}\n')
+
+        for currentType in typesByNamespace[namespace]:
+            print(f'### {currentType["Name"]}\n')
```
Comment (PunkPun) [inline packaging/format-docs.py]: On release there isn't a "\n" between name and description. Description also isn't formatted. I'm not sure if we need to change it

File: packaging/format-docs.py [hunk 13]
```diff
@@ -0,0 +1,82 @@
...
+        if currentType["Namespace"] in typesByNamespace:
+            typesByNamespace[currentType["Namespace"]].append(currentType)
+        else:
+            typesByNamespace[currentType["Namespace"]] = [currentType]
+
+    explanation = ""
+    if collectionName == "TraitInfos":
+        explanation = "all traits with their properties and their default values plus developer commentary"
+    elif collectionName == "WeaponTypes":
+        explanation = "a template for weapon definitions as well as its contained types (warheads and projectiles) with default values and developer commentary"
```
Comment (PunkPun) [inline packaging/format-docs.py]: Perhaps move `are` up to `explanations` so this part could use `is`

File: packaging/format-docs.py [hunk 14]
```diff
@@ -0,0 +1,82 @@
...
+                print(f'#### {currentType["Description"]}\n')
+
+            if "InheritedTypes" in currentType and currentType["InheritedTypes"]:
+                print("Inherits from: " + ", ".join([format_type_name(x) for x in currentType["InheritedTypes"]]) + '.\n')
+
+            if "RequiresTraits" in currentType and currentType["RequiresTraits"]:
+                print("Requires trait(s): " + ", ".join([format_type_name(x) for x in currentType["RequiresTraits"]]) + '.\n')
+
+            if len(currentType["Properties"]) > 0:
+                hasAttributes = "OtherAttributes" in currentType["Properties"][0]
```
Comment (PunkPun) [inline packaging/format-docs.py]: This is always true. See `TiberianSunRefinery`, it doesn't have any attributes on any of its values. 

This also fails at a lower part of the code

File: OpenRA.Mods.Common/UtilityCommands/ExtractWeaponDocsCommand.cs [hunk 15]
```diff
@@ -26,7 +28,7 @@ bool IUtilityCommand.ValidateArguments(string[] args)
 			return true;
 		}
 
-		[Desc("[VERSION]", "Generate weaponry documentation in MarkDown format.")]
+		[Desc("[VERSION]", "Generate weaponry documentation in JSON format.")]
```
Comment (PunkPun) [inline OpenRA.Mods.Common/UtilityCommands/ExtractWeaponDocsCommand.cs:31]: perhaps weaponry -> weapon

**Review by PunkPun — COMMENTED** [review]
Other than that LGTM

File: packaging/format-docs.py (hunk 11 above)
Comment (penev92) [reply packaging/format-docs.py]: I'm not trying to perfectly match the current format. I've always felt it makes it hard to tell where one entry starts and the other one ends, so I added a separator here.

File: packaging/format-docs.py (hunk 12 above)
Comment (penev92) [reply packaging/format-docs.py]: ~~I added formatting (`####`) to the description to make it stand out a tiny bit over the data in the table below it (and other text between the two). The empty line here is only so that the `####` actually works (it requires an empty line above it).~~
Changed the description to use `**...**` for bolding (hope it works?), which makes the empty line redundant now.

File: packaging/format-docs.py (hunk 13 above)
Comment (penev92) [reply packaging/format-docs.py]: But the current result is
> **"Listed below are a template for weapon definitions as well as its contained types (warheads and projectiles)"**

which I will probably change to
> **"Listed below are a template for weapon definitions and the types it can use (warheads and projectiles)"**

The `are` here sounds fine to me, as there are 2 things - the template and the list of types 🤔

File: packaging/format-docs.py (hunk 14 above)
Comment (penev92) [reply packaging/format-docs.py]: This is actually here to allow us to make a distinction between Trait docs, Weapon docs and Sequence docs, because the first two include attributes while the third one doesn't and we want to reuse the "table generation" code. I really should add a comment explaining that though because it did take me a while to figure out what you mean.

File: OpenRA.Mods.Common/UtilityCommands/ExtractWeaponDocsCommand.cs (hunk 15 above)
Comment (penev92) [reply OpenRA.Mods.Common/UtilityCommands/ExtractWeaponDocsCommand.cs:31]: I think `weaponry` is more fitting though 🤔

File: packaging/format-docs.py (hunk 11 above)
Comment (PunkPun) [reply packaging/format-docs.py]: I doubt it would look good on the website

Comment (penev92) [reply packaging/format-docs.py]: It is unfortunate that I can't easily test how it would look there, but IMO it looks reasonable with GitHub's markdown rendering 🤷‍♂️

File: packaging/format-docs.py (hunk 14 above)
Comment (PunkPun) [reply packaging/format-docs.py]: ah ok, then a comment is definitely needed

File: OpenRA.Mods.Common/UtilityCommands/ExtractWeaponDocsCommand.cs (hunk 15 above)
Comment (PunkPun) [reply OpenRA.Mods.Common/UtilityCommands/ExtractWeaponDocsCommand.cs:31]: Other than it sounding a bit more fancy I don't think it does. Better to keep naming consistent

**Comment by penev92** [comment]
Updated.

File: packaging/format-docs.py (hunk 12 above)
Comment (PunkPun) [reply packaging/format-docs.py]: `**...**` is literally the same as `####`

**Review by PunkPun — COMMENTED** [review]
I have a few styling suggestions

You should add a new line before every title. It looks a bit weird when reading a raw file that right after the table the `###` title starts

Screenshot 2022-08-22 at 19 47 37

I think we should consider not linking abstract classes as we don't actually list them here. \`...\` should also be removed from the as it looks better

Screenshot_2022-08-22_at_20 07 27

Perhaps adding `######` is also worth it. It makes the text light grey and reduces text size 

Screenshot_2022-08-22_at_20 08 51

I think `Trait` and `Warhead` classes should be blacklisted from the docs. They don't add anything of value, it doesn't make sense to mention that a trait inherits trait in every trait

Lastly markdown gives way too much space to attributes, it pushes descriptions rightways quite a lot. It could be fixed by inserting a space

Screenshot_2022-08-22_at_19 39 13

**Comment by penev92** [comment]
Updated documentation formatting. Here is the diff for easier verification:
<details>
  <summary>Click me</summary>
  
```diff
index 5bdd25d424..bd8b217885 100644
--- a/packaging/format-docs.py
+++ b/packaging/format-docs.py
@@ -11,12 +11,20 @@ import sys
 import json
 from collections import OrderedDict
 
-def format_type_name(typeName):
+def format_type_name(typeName, isKnownType):
     name = typeName
     if name.endswith("Info"):
         name = name[0:-4]
 
-    return f'[`{name}`](#{name.lower()})'
+    return f'[`{name}`](#{name.lower()})' if isKnownType else f'`{name}`'
+
+def is_known_type(typeName, types):
+    name = typeName
+    if name.endswith("Info"):
+        name = name[0:-4]
+
+    result = [t for t in types if name == t["Name"]]
+    return len(result) > 0
 
 def format_docs(version, collectionName, types):
     typesByNamespace = OrderedDict()
@@ -40,19 +48,21 @@ def format_docs(version, collectionName, types):
     print(f"Listed below are {explanation}.")
 
     for namespace in typesByNamespace:
-        print(f'\n## {namespace}\n')
+        print(f'\n## {namespace}')
 
         for currentType in typesByNamespace[namespace]:
-            print(f'### {currentType["Name"]}')
+            print(f'\n### {currentType["Name"]}')
 
             if currentType["Description"]:
-                print(f'**{currentType["Description"]}**\n')
+                print(f'**{currentType["Description"]}**')
 
             if "InheritedTypes" in currentType and currentType["InheritedTypes"]:
-                print("Inherits from: " + ", ".join([format_type_name(x) for x in currentType["InheritedTypes"]]) + '.\n')
+                inheritedTypes = [t for t in currentType["InheritedTypes"] if t not in ['TraitInfo', 'Warhead']] # Remove blacklisted types.
+                if inheritedTypes:
+                    print("###### Inherits from: " + ", ".join([format_type_name(x, is_known_type(x, types)) for x in inheritedTypes]) + '.')
 
             if "RequiresTraits" in currentType and currentType["RequiresTraits"]:
-                print("Requires trait(s): " + ", ".join([format_type_name(x) for x in currentType["RequiresTraits"]]) + '.\n')
+                print("###### Requires trait(s): " + ", ".join([format_type_name(x, is_known_type(x, types)) for x in currentType["RequiresTraits"]]) + '.')
 
             if len(currentType["Properties"]) > 0:
 
@@ -61,12 +71,13 @@ def format_docs(version, collectionName, types):
                 # While this can support per-entity hasAttributes, the use-case is per-file/document.
                 hasAttributes = "OtherAttributes" in currentType["Properties"][0]
 
+                print()
                 print(f'| Property | Default Value | Type |{" Attributes |" if hasAttributes else ""} Description |')
                 print(f'| -------- | ------------- | ---- |{" ---------- |" if hasAttributes else ""} ----------- |')
 
                 for prop in currentType["Properties"]:
                     if "OtherAttributes" in prop:
-                        attributesList = [f'{x["Name"]}{"(" + str(x["Value"][0]) + ")" if x["Value"] else ""}' for x in prop["OtherAttributes"] if x["Name"]]
+                        attributesList = [f'{x["Name"]}{" (" + str(x["Value"][0]) + ")" if x["Value"] else ""}' for x in prop["OtherAttributes"] if x["Name"]]
                         if not attributesList:
                             attributes = ''
                         else:
```
</details>

**Comment by penev92** [comment]
Fixed handling of custom attribute constructor parameters.

File: OpenRA.Mods.Common/UtilityCommands/ExtractTraitDocsCommand.cs [hunk 16]
```diff
@@ -35,84 +37,72 @@ void IUtilityCommand.Run(Utility utility, string[] args)
...
-				}
-			}
+					type.Namespace,
+					Name = type.Name.EndsWith("Info") ? type.Name.Substring(0, type.Name.Length - 4) : type.Name,
+					Description = string.Join(" ", type.GetCustomAttributes<DescAttribute>(false).SelectMany(d => d.Lines)),
+					RequiresTraits = RequiredTraitTypes(type)
+						.Select(y => y.Name),
+					InheritedTypes = type.BaseTypes()
+						.Select(y => y.Name)
+						.Where(y => y != type.Name && y != $"{type.Name}Info" && y != "Object" && y != "TraitInfo`1"),
```
Comment (PunkPun) [inline OpenRA.Mods.Common/UtilityCommands/ExtractTraitDocsCommand.cs]: ```y != "TraitInfo`1"``` feels like such a massive hack

Comment (penev92) [reply OpenRA.Mods.Common/UtilityCommands/ExtractTraitDocsCommand.cs]: Yes.

**Comment by PunkPun** [comment]
Weapons docks diff https://github.com/PunkPun/OpenRA/wiki/Weapons/_compare/7b725a8fc632b39859441de45fc5994a3c015523...04f4b99aef090f4936090949a5638ebeaa8f4051

It seems worse doesn't it?

**Comment by penev92** [comment]
No, not really 🤔

**Comment by PunkPun** [comment]
> No, not really 🤔

Well, I don't think that this is needed

Screenshot 2022-08-27 at 18 51 27

**Comment by penev92** [comment]
- It is needed
- It does give potentially useful info
- It doesn't get in anyone's way
- It doesn't interfere with anything
- I could make the parameters _italic_
- I'm quite done bike shedding over documentation formatting at this point

File: OpenRA.Mods.Common/UtilityCommands/ExtractTraitDocsCommand.cs (hunk 16 above)
Comment (PunkPun) [reply OpenRA.Mods.Common/UtilityCommands/ExtractTraitDocsCommand.cs]: Maybe add a comment explaining that it's a hack and why it is needed

**Review by PunkPun — COMMENTED** [review]
I'm unsure about the permissions of `format-docs.py`, on unix to run it I need to elevate them, won't this be a problem with github actions as well? And even if not I think it would be of benefit to just have permissions elevated for running the script from other sources

LGTM otherwise

**Comment by penev92** [comment]
Updated Python script execution based on help from @pchote.

**Review by PunkPun — DISMISSED** [review]
That fixed my permissions problem

**Comment by Mailaender** [comment]
> Well, I don't think that this is needed

I agree.

**Comment by penev92** [comment]
See, this is why we can't have nice things.

**Comment by Mailaender** [comment]
It dumps source code excerpts the average moddern can't understand, and it will overflow the already quite large width of the table.

**Comment by penev92** [comment]
The exact same thing can be said about the `Type` column. By the way, most of the `Default Value` column is empty, just taking up space and not all of the values in there are very clear either.

**Comment by Mailaender** [comment]
That one is more important for modding. Otherwise, you would be surprised by defaults. Especially since some still contain C&C actor names and stuff.

**Comment by Mailaender** [comment]
You could use

> [hover text](## "your hover text") for example.

or

> [id1]: ## "your hover text"
>
> This is another [hover text][id1] example with references.

but I don't know if @readthedocs supports it.

File: OpenRA.Mods.Common/UtilityCommands/ExtractTraitDocsCommand.cs (hunk 16 above)
Comment (PunkPun) [reply OpenRA.Mods.Common/UtilityCommands/ExtractTraitDocsCommand.cs]: ping

**Comment by penev92** [comment]
After a bunch more discussion[ on Discord](https://discord.com/channels/153649279762694144/388282819371204608/1015244158207807488) this is what I am settling on:

> a "compromise" solution - don't show the extra info in the docs, "encouraging" people to get and use ORAIDE to show them the rest 😏 

The "extra info" being the attribute's parameters. Now only showing attribute names.

Updated both example wiki pages (traits & weapons) to match the current state of the PR.

**Review by PunkPun — DISMISSED** [review]
:thumbsup:

**Comment by penev92** [comment]
The final result of all the generating and transforming can be seen here:
https://docs.openra.net/en/test/dev/traits/
https://docs.openra.net/en/test/dev/weapons/

File: packaging/format-docs.py [hunk 17]
```diff
@@ -0,0 +1,94 @@
...
+                    if "OtherAttributes" in prop:
+                        attributes = []
+                        for attribute in prop["OtherAttributes"]:
+                            attributes.append(attribute["Name"])
+
+                        defaultValue = ''
+                        if prop["DefaultValue"]:
+                            defaultValue = prop["DefaultValue"]
+                        elif 'Require' in attributes:
+                            defaultValue = '(require)'
```
Comment (PunkPun) [inline packaging/format-docs.py]:
```suggestion
                            defaultValue = '(required)'
```

**Review by PunkPun — DISMISSED** [review]

**Comment by Mailaender** [comment]
[Changelog](https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/1e32b0fa4cf87096445217d6e6d3d93c9658ed5c...34201aab2b06d0788ac230d40986de3596231102)
